### PR TITLE
Implement Debug trait for some public types

### PR DESCRIPTION
**Pull request by PolyMeilex**
List of reviewed modules/files:
- [ ] smithay
    - [ ] backend
        - [ ] drm
            - [x] atomic
            - [x] common
            - [x] egl
            - [x] eglstream
            - [ ] gbm
            - [ ] legacy
            - [x] mod.rs
        - [x] egl
        - [x] graphics
        - [x] libinput
        - [x] session
        - [x] input.rs
        - [x] mod.rs
        - [x] udev.rs
        - [x] winit.rs
    - [x] utils
    - [x] wayland
    - [x] xwayland
    - [x] lib.rs
    - [x] reexports.rs
    - [x] signaling.rs

### CODE REVIEW

**Comment by psychon** [comment]
Would it make sense to enable warnings on the `missing-debug-implementations` lint? Otherwise, this issue will likely come up again and again.

**Comment by PolyMeilex** [comment]
> Would it make sense to enable warnings on the `missing-debug-implementations` lint? Otherwise, this issue will likely come up again and again.

In some cases implementing Debug does not make too much sense, but still I would advocate for `missing-debug-implementations` to be enabled by default, so it would have to be explicitly disabled every time you think Debug is useless.
TLDR: I think implementing Debug should be opt-out instead of opt-in

**Comment by vberger** [comment]
Ah yes, it should definitely be done in calloop as well you're right.

**Comment by PolyMeilex** [comment]
With those 2 commits I think I'm done with this PR, as I'm not too keen on working on remaining `drm/gbm` and `drm/legacy` Debug traits.
According to `missing-debug-implementations` lint, after this PR most of the crate is covered (with only a few exceptions)

**Comment by vberger** [comment]
Hi, thanks for this PR, and sorry for the late answer.

It looks pretty good overall, but in the name of future maintenance, I think it'd be nice to add a comment explaining why a manual debug impl was necessary when it was. Because some of them are not obvious to me, as they seem to be mostly doing exactly what `derive(Debug)` would have done...

Also, looks like this PR needs to be rebased, (#255 removed the `xwayland/launch_helper.rs` file).

**Comment by PolyMeilex** [comment]
Sure, I'll review all of them once again, to make sure that there are no unnecessary manual impls left over, and I'll add comments as I go.

**Comment by PolyMeilex** [comment]
Oh good, I believe that I did not cause those build errors, it looks like it cheapens on master branch as well. 
Looks like `nix` crate does not follow semver properly. (or maybe I'm wrong, no idea)

**Comment by vberger** [comment]
Agh, again a `nix` update. This is #229 yet again. I'll update nix on the repo, so that you can rebase on top of that.

**Comment by vberger** [comment]
Okay, I've done a pass of dependencies-updating, that should be good now.

**Comment by PolyMeilex** [comment]
Done, I believe that I removed all unnecessary manual impls (sorry about that, It happened because I did not add all of them in one go, and some of the impls became obsolete after the whole PR was done, and I just did not notice), and the PR is successfully rebased onto master.

File: src/backend/libinput/mod.rs [hunk 1]
```diff
@@ -423,6 +427,26 @@ impl From<event::pointer::ButtonState> for backend::MouseButtonState {
...
+                    .finish()
+            }
+        }
+
+        f.debug_tuple("LibinputSessionInterface")
+            .field(&SessionDebug(&self.0))
+            .finish()
+    }
+}
+
```
Comment (vberger) [inline src/backend/libinput/mod.rs:449]: Same here

File: src/backend/session/direct.rs [hunk 2]
```diff
@@ -160,6 +162,25 @@ pub struct DirectSessionNotifier {
...
+                "source",
+                &match self.source {
+                    Some(_) => "Some(..)",
+                    None => "None",
+                },
+            )
+            .finish()
+    }
+}
+
```
Comment (vberger) [inline src/backend/session/direct.rs:183]: This one is missing a comment, I guess it's because `Signals` does not implement `Debug`?

File: src/backend/drm/common/fallback.rs [hunk 3]
```diff
@@ -49,6 +50,22 @@ pub enum FallbackDevice<D1: Device + 'static, D2: Device + 'static> {
...
+                .field("device_id", &d.device_id())
+                .finish(),
+            FallbackDevice::Fallback(d) => f
+                .debug_struct("FallbackDevice::Preference")
+                .field("device_id", &d.device_id())
+                .finish(),
+        }
+    }
+}
+
```
[the remark this answers is not in the document]
Comment (PolyMeilex) [reply src/backend/drm/common/fallback.rs:68]: Sure, I'll stick to derive than

File: src/backend/session/direct.rs (hunk 2 above)
Comment (PolyMeilex) [reply src/backend/session/direct.rs:183]: yep, I added a comment for that

**Review by vberger — APPROVED** [review]

**Comment by vberger** [comment]
Thank you!
